datafusion/core/tests/sqllogictests/src/insert/mod.rs

```diff
@@ -0,0 +1,96 @@
+// under the License.
+
+mod util;
+
+use crate::error::{DFSqlLogicTestError, Result};
+use crate::insert::util::LogicTestContextProvider;
+use datafusion::datasource::MemTable;
+use datafusion::prelude::SessionContext;
+use datafusion_common::{DFSchema, DataFusionError};
+use datafusion_expr::Expr as DFExpr;
+use datafusion_sql::parser::{DFParser, Statement};
+use datafusion_sql::planner::SqlToRel;
+use sqlparser::ast::{Expr, SetExpr, Statement as SQLStatement};
+use std::collections::HashMap;
+
+pub async fn insert(ctx: &SessionContext, sql: String) -> Result<String> {
+    // First, use sqlparser to get table name and insert values
+    let mut table_name = "".to_string();
+    let mut insert_values: Vec<Vec<Expr>> = vec![];
+    if let Statement::Statement(statement) = &DFParser::parse_sql(&sql)?[0] {
+        if let SQLStatement::Insert {
+            table_name: name,
+            source,
+            ..
+        } = &**statement
+        {
+            // Todo: check columns match table schema
+            table_name = name.to_string();
+            match &*source.body {
+                SetExpr::Values(values) => {
+                    insert_values = values.0.clone();
+                }
+                _ => {
+                    return Err(DFSqlLogicTestError::NotImplemented(
+                        "Only support insert values".to_string(),
+                    ));
+                }
+            }
+        }
+    } else {
+        return Err(DFSqlLogicTestError::Internal(format!(
+            "{:?} not an insert statement",
+            sql
+        )));
+    }
+
+    // Second, get table by table name
+    // Here we assume table must be in memory table.
+    let table_provider = ctx.table_provider(table_name.as_str())?;
+    let table_batches = table_provider
+        .as_any()
+        .downcast_ref::<MemTable>()
+        .ok_or_else(|| {
+            DFSqlLogicTestError::NotImplemented(
+                "only support use memory table in logictest".to_string(),
+            )
+        })?
+        .get_batches();
+
+    // Third, transfer insert values to `RecordBatch`
+    // Attention: schema info can be ignored. (insert values don't contain schema info)
+    let sql_to_rel = SqlToRel::new(&LogicTestContextProvider {});
+    let mut insert_batches = Vec::with_capacity(insert_values.len());
+    for row in insert_values.into_iter() {
+        let logical_exprs = row
+            .into_iter()
+            .map(|expr| {
+                sql_to_rel.sql_to_rex(expr, &DFSchema::empty(), &mut HashMap::new())
+            })
+            .collect::<std::result::Result<Vec<DFExpr>, DataFusionError>>()?;
+        // Directly use `select` to get `RecordBatch`
+        let dataframe = ctx.read_empty()?;
+        insert_batches.push(dataframe.select(logical_exprs)?.collect().await?)
+    }
+
+    // Final, append the `RecordBatch` to memtable's batches
+    let mut table_batches = table_batches.write();
```

Comment:
Rather than changing the batches in the existing memtable, what would you think about creating a new memtable with the same name with the new values (rather than modifying the original one)



I think you might be able to avoid changes to SessionContext and MemTable entirely.

Something like this (untested)

```rust
// fetch existing batches
let mut existing_batches = ctx.table(table_name.as_str()).collect();
// append new batch
exsiting_batches.extend(insert_batches)

// Replace table provider provider
let new_provider = MemTable::try_new(batches[0].schema(), vec![batches]);
ctx.register_table(table_name, new_provider)
```

Reply:
Yes, I have thought about the way you mentioned, (also need to delete original memtable).

-- But for performance reasons, I choose the current way: modifying the original one.

If you think the changes to add interior mutability to MemTable don't make sense, I can change it in the following ticket! (I don't have a strong bias for both ways)

Reply:
I think keeping MemTable as simple as possible is likely the best approach -- so for that reason I prefer to remove the interior mutability.

I can give it a shot if you agree -- I think the performance of copying record batches (for reasonably small data such as what is in the test) will be ok

Reply:
Ok, agree! Thanks @alamb . I'll refactor it later.

Reply:
Thank you very much!